Health visual effect

ChillBeanJosh opened the pull request:
-Updated slightly so lower health seems more noticeable.

-Just was the addition of bloom able to be toggled off as well.

ChillBeanJosh committed:
Health Visual Effect Implementation.

-Created a Health Visual Effect Controller which uses Post Processing to add Visuals portraying the players health having the intensity increase the lower the players health gets.

-Attach Player Killable to killable and will work

-Also added toggles to toggle on and off some effects to see how they look with and without.

-Saved as a prefab to be easily moved over to other scenes

Note: On default Post Processing wasnt turned on, needed to be toggled on the camera from the Rendering section so if nothing is appearing that may be the cause.

ChillBeanJosh committed:
Health Visual Effect

ChillBeanJosh committed:
Slight mistake fix

-Fixed to make the bloom work as i was referencing the threshold instead of the intensity.

-Apologies for the multiple commits.

TheStress commented on Assets/Scripts/UI/HealthEffectController.cs:
```diff
@@ -0,0 +1,134 @@
+   void UpdateVisualEffect(float health)
+   {
+
+    float intensityController = Mathf.Clamp01(1.0f - (health / 100.0f));
```
For each of these effects it should be a percentage of a value in between two values. For each of them they should have a start and max value, then calculate a value in between using the percentage of the health, like `currentHealth/maxHealth`. 

So the code would look like this:
`float newValue = Mathf.lerp(startValue, maxValue, currentHealth/maxHealth)`

Then you should apply the value using `Override()` since it allows you to go above the max 1 instensity, like `vignette.intensity.Override(newValue)`

TheStress commented on Assets/Scripts/UI/HealthEffectController.cs:
```diff
@@ -0,0 +1,134 @@
+
+    if(volume.profile.TryGet<FilmGrain>(out filmgrain))
+    {
+        killable.OnHealthChange.AddListener(OnHealthChange);
```
you only need to add this listener once

TheStress commented on Assets/Scripts/UI/HealthEffectController.cs:
```diff
@@ -0,0 +1,134 @@
+   void Start()
+   {
+
+    if(volume.profile.TryGet<Vignette>(out vignette))
```
It looks like you arent setting any of the components you are getting here to the variables you declare above

TheStress commented on Assets/Scripts/UI/HealthEffectController.cs:
```diff
@@ -0,0 +1,134 @@
+    }
+    else
+    {
+        vignette.intensity.value = 0.0f;
```
with the implementation I suggested above you dont set the values to 0 because we want some of the post processing

TheStress reviewed (commented):
Changes requested in the code

ChillBeanJosh committed:
Updated Request Changes

- Changes how values work to be anything by being given a start and max which can all be changed and adjusted in the inspector.

- Removed other copies of listener so it is added once.

- Adjusted the else to when a bool is toggled off, becomes the start value in the case the effect always wants to be portrayed.

TheStress commented on Assets/Scripts/UI/HealthEffectController.cs:
```diff
@@ -0,0 +1,167 @@
+    }
+    else
+    {
+        filmgrain.intensity.Override(filmgrainStart);
```
you dont need these else atatements since you only apply them with the toggle

TheStress commented on Assets/Scripts/UI/HealthEffectController.cs:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    //EFFECTS FILM GRAIN
+    if(filmgrainToggle)
```
In addition to the toggle you should also check if each respective component != null

TheStress commented on Assets/Scripts/UI/HealthEffectController.cs:
```diff
@@ -0,0 +1,167 @@
+    if(volume.profile.TryGet<Vignette>(out vignette))
+    {
+        vignetteToggle = true;
+        killable.OnHealthChange.AddListener(OnHealthChange);
```
you should take the add listener and put it outside the if statement

TheStress commented:
You should also add chromatic aberration

ChillBeanJosh committed:
Health Effect Touchup

- Added != null in addition to the toggle

- Moved the listener outside the if statement.

- Added Chromatic Aberration as another adjustable setting.

TheStress committed:
adjust variables + clean up code
